docs: add new consumer concept doc (#991)
asyncapi/website: pages/docs/concepts/consumer.md

---
title: Consumer
weight: 4
---

## What is a Consumer?

> In an Event Driven Architecture (EDA), a consumer is a component of an application or messaging system that listens for a particular event from a broker and takes the necessary action in response.

Reviewer: ```suggestion
In an Event Driven Architecture (EDA), a consumer is an application that listens for a particular event from a broker and reacts to it.
```

> application

I think it is good to be consistent with producer document that was just released

> takes the necessary action in response

I would avoid `response` as it is known term in request/response pattern, and might be suggest consumer sends a response

> ## Why do we need Consumers?

Reviewer 2: ```suggestion
## What is the purpose of Consumers?
```

This is the outline all **Concept** Docs will have: `What is the purpose of _____?`

Author: Okay, I will correct that asap!

Reviewer 2: don't hate but i changed my mind :D haha!

Let's stick with your original one, I went ahead and fixed it :D

Author: Alrighty!

> Unlike traditional RestAPIs, EDA creates a setting where there is never a wait for requests or responses to an event. In this pattern, a consumer is unaware of the producer or other consumers; all they know is that when a broker sends them an event they are subscribed to, they receive it.

Reviewer: ```suggestion
Unlike traditional REST APIs, in EDA, consumers of the event are not expected to respond immediately on the same connection. In this architecture, a consumer is unaware of the producer or other consumers; all they know is that when a broker sends them an event, it is because they subscribed to it.
```

how about this? just to clarify about communication that it is not on sync, because sometimes people do request/response in EDA, but it is still async. Like in DM on Slack, you send question, you do not know who will answer, but you definitely expect that in a thread at some point of time someone will send a response. Makes sense

you can freely adjust the paragraph, my suggestions are just to "visualize" what I mean

When you want events processed asynchronously in your application, the consumer plays an important role in completing that flow of event data in the event channel.

```mermaid
flowchart LR
    a[Producer] -- Event A --->c[(Broker)]
    a -- Event B ---> c
    c -- Event A ---> d[Consumer A]
    c -- Event A ---> e[Consumer B]
    c -- Event B ---> e[Consumer B]
    c -- Event B ---> f[Consumer C]
    subgraph Consumers
    d
    e
    f
    end
```

> The diagram above depicts an example of the flow of events from a producer to a broker and then to the consumers. In this instance, the producer publishes two events, A & B and sends to the broker. Then each consumer receives events from what it is subscribed to.

Reviewer: ```suggestion
The diagram above depicts an example of the flow of events from a producer to a broker and then to the consumers. In this instance, the producer publishes two events, A & B, and sends them to the broker. Then each consumer receives events that they are subscribed to.
```

weight: 4
---

## What is a consumer?
In an Event Driven Architecture (EDA), a consumer is an application that listens for a particular event from a broker and reacts to it.

## Why do we need consumers?
Unlike traditional REST APIs, in EDA, event consumers are not expected to respond immediately on the same connection. In this architecture, a consumer is unaware of the producer or other consumers; all they know is that when a broker sends them an event, it is because they subscribed to it.

When you want events processed asynchronously in your application, the consumer plays an important role in completing that event data flow in the event channel.

```mermaid
flowchart LR
    a[Producer] -- Event A --->c[(Broker)]
    a -- Event B ---> c
    c -- Event A ---> d[Consumer A]
    c -- Event A ---> e[Consumer B]
    c -- Event B ---> e[Consumer B]
    c -- Event B ---> f[Consumer C]
    subgraph Consumers
    d
    e
    f
    end
```
The above diagram depicts a sample flow of events from producer to broker and consumer. In this instance, the producer publishes two events, A & B, and sends them to the broker. Then each consumer subscribes to receive those events.

<Remember>

> `Subscribers` can also be [`producers`](https://www.asyncapi.com/docs/concepts/producer).

Reviewer: I don't think markdown render well in the <Remember>  component

<img width="2103" alt="Screenshot 2022-10-12 at 09 52 08" src="https://user-images.githubusercontent.com/6995927/195283179-2696157c-06da-450e-b95f-723a9b2cd394.png">

isn't it better to just do markdown note/quote `>`

Reviewer 2: Ahhh haha noooo. ✌🏽😂

In this scenario, I believe Anisat's original idea to use our React `Remember` component is a good idea because it's a helpful visual reminder that stands out. 

The solution (to me) is to instead use appropriate HTML tags within our React `Remember` component, and then teach Anisat that we can't use MD syntax in these components. 😄🙌🏽

I fixed it and think this looks good to merge now too 🎉🎉🎉
<img width="800" alt="Screen Shot 2022-10-24 at 5 13 42 PM" src="https://user-images.githubusercontent.com/19964402/197653523-7b2d2a85-9c08-4172-ba60-7b7852750695.png">
